### DISCUSSION TITLE
How to use GraphQL queried data in SSR rendered `html.js`?

### DISCUSSION THREAD

**Discussion by paulwellnerbou** [gatsbyjs/gatsby#38930] [Help] [answered]
We are using Strapi as CMS, and Gatsby as static site generator, all HTMLs, JS, CSS, images and other resources will be packaged and deployed independently from the original CMS, as a static site.

We use React and GraphQL static queries already for client components.

Now I want to add custom (global) stylesheets which can be maintained in a CMS (Strapi) which is renderend on build time into the HTML. We have our custom `html.js`:

```jsx
import React from "react"
import PropTypes from "prop-types"
import SsrCustomStyles from "./ssr/SsrCustomStyles"

export default function HTML(props) {
  return (
    <html {...props.htmlAttributes}>
      <head>
        ...
        {props.headComponents}
        <SsrCustomStyles />
      </head>
      <body {...props.bodyAttributes}>
     ....
```

In `SsrCustomStyles` I am trying to get data from GraphQL, but the query result is always empty. The query and my logic works, I was able to make it work with `gatsby develop`.

How do I fetch data with GraphQL at build time in a component included in `html.js`?

What I tried:

1. just defining a query in this file and expecting `{ data }` to be passed automatically as prop, as documented here: https://www.gatsbyjs.com/docs/reference/rendering-options/server-side-rendering/#interplay-with-build-time-graphql-queries
    
```
  WebpackError: TypeError: Cannot read properties of undefined (reading 'allStrapiAppSettings')
```

Variable `data` is always empty.

2. `useStaticQuery`

Doesn't build, I get following error:
    
```
 ERROR  UNKNOWN

Truncated page data information for the failed page "/offline-plugin-app-shell-fallback/": {
  "errors": {},
  "path": "/offline-plugin-app-shell-fallback/",
  "slicesMap": {},
  "pageContext": {}
}

failed Building static HTML for pages - 2.622s

 ERROR #95313  HTML.COMPILATION

Building static HTML failed for path "/offline-plugin-app-shell-fallback/"

See our docs page for more info on this error: https://gatsby.dev/debug-html

  81 |     return context[query].data
  82 |   } else {
> 83 |     throw new Error(
     |           ^
  84 |       `The result of this StaticQuery could not be fetched.\n\n` +
  85 |         `This is likely a bug in Gatsby and if refreshing the page does not fix it, ` +
  86 |         `please open an issue in https://github.com/gatsbyjs/gatsby/issues`

  WebpackError: The result of this StaticQuery could not be fetched.
```

3. <StaticQuery ...>
    Does not work, in the final HTML there is just a `<div>Loading (StaticQuery)</div>`

An example for the latter case:

```jsx
export default function SsrCustomStyles({ data }) {
    return (
      <StaticQuery query={query} render={data => {
          const settings = data.allStrapiAppSettings.nodes[0]
          const files = data.allCssFiles.nodes[0]
          return <>
              {getStyleSheets(settings, files).map(
                ({ id, href }) => <StyleSheet key={id} id={id} href={href} />
              )}
          </> }
      } />
    )
}

export const query = graphql`
    query {
        allStrapiAppSettings { nodes { ... } }
        allCssFiles: allFile(filter: {extension: {in: ["css"]}}) { nodes { ... } }
    }
`
```

**Comment by violy** [comment]
Hi Paul, 

Using `html.js` is **probably not the best to add dynamic stylesheet**. 
The best solution might be to use the [Gatsby Head API](https://www.gatsbyjs.com/docs/reference/built-in-components/gatsby-head/) and [Gatsby Browser API](https://www.gatsbyjs.com/docs/reference/config-files/gatsby-browser/) with `wrapRootElement`.   

As mention in Gatsby Head Api Doc :
> As of gatsby@5.6.0, Head can access [React Context](https://reactjs.org/docs/context.html) that you defined in the [wrapRootElement API](https://www.gatsbyjs.com/docs/reference/config-files/gatsby-browser/#wrapRootElement). It’s important to note that wrapRootElement should only be used to set up context providers. UI components should be defined in [wrapPageElement API](https://www.gatsbyjs.com/docs/reference/config-files/gatsby-browser/#wrapPageElement).

I think this is the best option in Gatsby.

**Reply by paulwellnerbou** [reply]
@violy Thanks, for your answer. But that's not what I want. I did this already, and it kind of works, but it causes flickering, because the styles are added/changed via JavaScript in the browser.

I want it already in the resulting HTML.

So even if this is not the best option -- is there still a way to do it?

An alternative would be to generate the stylesheets into a `customstyles.css` file and reference this in `html.js` directly. But not sure how this is possible without any further hacks.

**Reply by violy** [reply]
Sorry, I didn't mention that `wrapRootElement` and `wrapPageElement` is exposed both by Gatsby **Browser** API and Gatsby **SSR** API – so you can use the same function / component twice. Have you already do that ? It would prevent the flickering… ?

**Reply by violy** [reply]
Also, I've seen a `<StyleSheet />` component with `href` prop in your code, what is that ? Is `href` value an absolute CSS url ? And so, why not just using standard `<link 
 rel="stylesheet" href={href} />` instead ?

**Reply by paulwellnerbou** [reply]
> ...wrapRootElement and wrapPageElement is exposed both by Gatsby Browser API and Gatsby SSR...

Thanks, I'll look into it and try again. The documentation you quoted:

> It’s important to note that wrapRootElement should only be used to set up context providers. UI components should be defined in [wrapPageElement API](https://www.gatsbyjs.com/docs/reference/config-files/gatsby-browser/#wrapPageElement).

That would be against this advice, correct?

> Also, I've seen a <StyleSheet /> component...

This is a component switching between `<link rel="stylesheet" ...>` and `<style> ... </style>` depending on the "type" of stylesheet. But yes, for our case we can just assume it is a standard `<link  rel="stylesheet" href={href} />`.

**Reply by violy** [reply, marked as answer]
Have you try to use Head Api ? 

I show you an example, using bootstrap css cdn, and a simple static query as a Proof of Concept…

First, create a `DefaultHead` component, like this
```Typescript
import React from "react";
import {graphql, HeadProps, useStaticQuery} from "gatsby";

export type DefaultHead<DataType = object, PageContextType = object> = (
    props: HeadProps<DataType, PageContextType> & { title?: string }
) => JSX.Element
const DefaultHead: DefaultHead<any,any> = (props) => {
    const { site: { siteMetadata } } = useStaticQuery(graphql`
        query DefaultHead {
            site {
                siteMetadata {
                    title
                }
            }
        }`);
    const { title = siteMetadata.title || "Default Title" } = props;
    return (<>
        <link href={"https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta1/dist/css/bootstrap.min.css"} rel={"stylesheet"} />
        <title>{title}</title>
    </>);
}
export default DefaultHead;
```

Then, in all of your page components, using Gatsby Head Api, add something like this without flickering in my browser…
```Typescript
export const Head: HeadFC = (props) => <DefaultHead {...props} />
```

in my static html generated at build time, I have following lines:  
```html
<link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta1/dist/css/bootstrap.min.css" rel="stylesheet" data-gatsby-head="true"/>
<title data-gatsby-head="true">my site title from GraphQL</title>
```

in this sample, the CSS url is static, but you can use any graphql result as source of data.

**Reply by paulwellnerbou** [reply]
@violy Thank you very much!

Trying your solution I had a deeper look at my project. I was already using `<Helmet>`, and in our builds, the stylesheets are already rendered in the resulting HTML files! :man_facepalming: 

The flickering (and not always correct rendering) only happens in the preview, when running `gatsby develop`.

So it works all fine, and I did not have to change anything.

Sorry for the confusion -- and sorry for the question at all. Nevertheless it helped me a ton to understand more about the rendering process, the APIs, and so on. So thank you so much! :pray:
